src/backend/functions/orchestrator/lib/index.js

```diff
@@ -102,35 +110,100 @@ function createDeleteParams({mediaUrl, tasksTableName}) {
+const terminated = R.curry(deleteItem => o(deleteItem, createDeleteParams));
 
-function error(ecs, ddb) {}
+// errorUpdate :: ({k: v} -> Promise {k: v}) -> {k: v} -> Promise {k: v}
+const errorUpdate = R.curry((updateItem, {tasksTableName, mediaUrl, taskArn}) => {
+   const params = {
+      TableName: tasksTableName,
+      Key: {
+         MediaUrl: mediaUrl
+      },
+      UpdateExpression: 'ADD Retries :val SET TaskStatus = :status, TaskArn = :task',
+      ExpressionAttributeValues: {
+         ':val': 1,
+         ':status':  'WAITING',
+         ':task': taskArn
+      },
+      ReturnValues: 'ALL_NEW'
+   };
+
+   return updateItem(params)
+       .then(({Attributes}) => ({...Attributes, tasksTableName}))
+});
+
+const unrecoverableError = R.curry((updateItem, {tasksTableName, mediaUrl}) => {
+   const params = {
+      TableName: tasksTableName,
+      Key: {
+         MediaUrl: mediaUrl
+      },
+      UpdateExpression: 'SET TaskStatus = :status',
+      ExpressionAttributeValues: {
+         ':status':  'UNRECOVERABLE_ERROR'
+      },
+      ReturnValues: 'ALL_NEW'
+   };
+
+   return updateItem(params);
+});
 
+// error :: AWS.ECS -> AWS.DynamoDB -> {k: v} -> Promise {k:v}
+const error = R.curry((ecs, updateItem) => composeP(errorUpdate(updateItem), startTranscription(ecs)));
 
-// handler :: AWS.ECS -> AWS.DynamoDB -> {k: v} -> ({k: v} -> {k: v} -> Promise {k: v})
+// handler :: AWS.ECS -> AWS.DynamoDB -> {k: v} -> ({k: v} -> Promise {k: v})
 module.exports = (ecs, ddb, env) => {
+   const update = params => ddb.update(params).promise();
+
    const handlers = {
-      WAITING: waiting(ecs, params => ddb.update(params).promise()),
+      WAITING: waiting(ecs, update),
       INITIALIZING: noop,
       PROCESSING: noop,
       TERMINATING: terminating(ecs),
       TERMINATED: terminated(params => ddb.delete(params).promise()),
-      ERROR: error
+      ERROR: error(ecs, update),
+      UNRECOVERABLE_ERROR: noop
    };
 
-   return (event, context) => {
+   const normalisedEnv = convertEnvVars(env);
+
+   return event => {
+
       const promises = R.map(record => {
          const image = converter.unmarshall(R.path(['dynamodb', 'NewImage'], record));
-         const handler = R.defaultTo(noop, handlers[image.TaskStatus]);
-         return handler(env, image);
+
+         const params = R.mergeRight(normalisedEnv, convertImage(image));
+
+         const {mediaUrl, retries, taskStatus, tasksTableName, retryThreshold} = params;
+
+         console.log('Status: ' + taskStatus);
+
+         if(R.defaultTo(0, retries) > parseInt(retryThreshold)) {
+            console.log('Error threshold exceeded');
+            return unrecoverableError(update, {mediaUrl, tasksTableName});
+         }
+
+         const handler = R.defaultTo(noop, handlers[taskStatus]);
+         return handler(params)
+             .catch(createEnrichedError({tasksTableName, mediaUrl}));
```

Comment:
I think it's right to call `createEnrichedError` here, but on L93 you throw an enriched error, which will result in the same log statement twice. maybe above we can just throw an ordinary error rather than calling enrich twice?

Reply:
The problem is I need to thread the `taskArn` so maybe what I should do above is just attach the `taskArn` and nothing else in the throw part.

Reply:
And perhaps the log statement shouldn't be there at all.

Reply:
yeah I agree, can move the logging elsewhere as long as it's logged somewhere. actually once you remove the log statement it makes sense to call it twice because you're enriching with different information which is fine. I think the name `createEnrichedError` threw me because in my mind it was making it akin to creating a custom error type, but the function is actually just enriching an error in an immutable manner - all good

Reply:
*done